### PR TITLE
docs: Document the hooks that Cilium uses

### PR DESCRIPTION
**Pull request by joestringer**

### CODE REVIEW

File: Documentation/internals/hooks.rst [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+eBPF Hooks
+==========
+
+Cilium uses the following eBPF hooks to attach eBPF programs to the kernel:
+
+- ``BPF_PROG_TYPE_XDP``
+- ``BPF_PROG_TYPE_SCHED_ACT``
+- ``BPF_PROG_TYPE_CGROUP_SOCK_ADDR``
+- ``BPF_PROG_TYPE_SOCK_OPS``
+- ``BPF_PROG_TYPE_SK_MSG``
```
Comment (dylandreimerink) [inline Documentation/internals/hooks.rst]: What do we mean by "hooks"? I would call these program types, some of which can attach at multiple locations in the kernel, take `BPF_PROG_TYPE_CGROUP_SOCK_ADDR` for example which we attach at `connect4`, `bind4`, ect.

Comment (joestringer) [reply Documentation/internals/hooks.rst]: Makes sense yeah. I'll update to define this as the program types. I don't plan on documenting all of the individual function attachment locations in this PR.

File: Documentation/internals/hooks.rst [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+eBPF Program Types
+==================
+
+Cilium uses the following eBPF program types to attach programs to the kernel:
+
+- ``BPF_PROG_TYPE_XDP``
+- ``BPF_PROG_TYPE_SCHED_ACT``
+- ``BPF_PROG_TYPE_CGROUP_SOCK_ADDR``
+- ``BPF_PROG_TYPE_SOCK_OPS``
```
Comment (aditighag) [inline Documentation/internals/hooks.rst]: I'm not sure if we use the `*SOCK_OPS` or `SK_MSG` hooks.

Comment (joestringer) [reply Documentation/internals/hooks.rst]: We still have sockmap code in the tree, so even if it's not always enabled, it is possible for Cilium to use these hooks.

File: Documentation/internals/hooks.rst [hunk 3]
```diff
@@ -0,0 +1,16 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    https://docs.cilium.io
+
+eBPF Program Types
+==================
```
Comment (qmonnet) [inline Documentation/internals/hooks.rst:8]: Suggestion: Make it more explicit from ToC that we're not documenting the types?

```suggestion
eBPF Program Types in Use
=========================
```

**Review by ldelossa — APPROVED** [review]
Welcome addition.

File: Documentation/internals/hooks.rst (hunk 2 above)
Comment (qmonnet) [reply Documentation/internals/hooks.rst]: About that, should there be a distinction between hooks that we're using all the time (Is there any?) and those that are feature-dependant?

File: Documentation/internals/hooks.rst (hunk 3 above)
Comment (joestringer) [reply Documentation/internals/hooks.rst:8]: Is this ambiguous? The text immediately below specifies that Cilium uses the following types. This is also only linked from the Cilium internals page.

Comment (qmonnet) [reply Documentation/internals/hooks.rst:8]: OK yeah, the text below it is probably good enough indeed

**Comment by qmonnet** [comment]
This PR is probably still of interest, I'm removing the `stale` label for this time.

**Comment by joestringer** [comment]
Given #23555 , I'm going to drop the documentation around sockops.

**Review by qmonnet — APPROVED** [review]
Thanks for this

**Pull request merged**